google-cloud-spanner/src/main/resources/com/google/cloud/spanner/spi/v1/grpc-gcp-apiconfig.json

```diff
@@ -0,0 +1,106 @@
+{
+  "channelPool": {
+    "maxSize": 3,
+    "maxConcurrentStreamsLowWatermark": 1
+  },
+  "method": [
+    {
+      "name": ["google.spanner.v1.Spanner/CreateSession"],
+      "affinity" : {
+        "command": "BIND",
+        "affinityKey": "name"
+      }
+    },
+    {
+      "name": ["google.spanner.v1.Spanner/BatchCreateSessions"],
```

Comment:
Correct, this will bind (create an affinity) the channel with the sessions created.
We do not use the `CHANNEL_HINT`

Yes, everything above will work with the new channel pool but in a slightly different manner.

When the new channel pool gets a request without a session (like BatchCreateSessions request) it picks the least busiest channel (i. e. the channel with the minimum number of in-flight RPCs).
Also if max number of channels is not reached and the least busiest channel has at least `maxConcurrentStreamsLowWatermark` in-flight RPCs then a new channel will be created and provided.

Thus having `maxConcurrentStreamsLowWatermark` set to 1 and having the first N requests to be `BatchCreateSessions` which are executed concurrently ensures that each of N channels will receive exactly one BatchCreateSessions request.

Reply:
Thanks for the detailed explanation. I think this will work in most cases, but I have a couple of questions / worries for some corner cases.

I worry a little bit about the way that a channel is picked for the `BatchCreateSessions` RPC. In most cases for the initial creation of the session pool, it will probably work well, as the initial calls are initiated in parallel. There are still a couple of scenarios that could cause the behavior to change from what it is currently:
* If one of the `BatchCreateSessions` invocations should finish before the background executor has picked up all of the requests, it could be that one channel is assigned more calls than the others. This is not very probable when using the default settings and using only one `DatabaseClient`, as the executor uses more threads than the default number of channels, and each channel only needs one invocation (one invocation of `BatchCreateSessions` will never return more than 100 sessions, even if the requested session count is more than 100). However, the [default executor](https://github.com/googleapis/java-spanner/blob/0f56261857aa9cb516bf31f796e1badb0a792c86/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerImpl.java#L181) that is used for the `SessionClient` is shared among all `DatabaseClient` instances that are created from a `Spanner` instance. That means that if an application creates multiple `DatabaseClient` instances (for different databases) during startup, the executor could be fully exhausted for a while, causing the initial `BatchCreateSessions` for at least some of the `DatabaseClient`s to not be executed in parallel.
* The default configuration sets `maxConcurrentStreamsLowWatermark` to 1 to force the creation of a new channel each time a (parallel) request is done. Is that only used to determine when to create a new channel in the pool, or will it also affect the number of parallel streams on a channel that can be used during the lifetime of the application?
* What happens if the user sets a higher value for `maxConcurrentStreamsLowWatermark` in their own configuration? (Can they do that?)

Reply:
Thank you for pointing this out! Yeah, if there is a chance of some BatchCreateSession requests to finish before some others are started during the session pool creation, then we can go safer with configuring `maxConcurrentStreamsLowWatermark` to `0`. In this case a new channel will be created for each BatchCreateSession request regardless of if they are running concurrently or sequentially.

Answering two other questions:

- `maxConcurrentStreamsLowWatermark` is only used to determine if a new channel should be created. And it works up until the point when the pool reaches the maximum size.
- A user cannot change the `maxConcurrentStreamsLowWatermark`. `maxConcurrentStreamsLowWatermark` is configured via `ApiConfig` (together with affinity configuration for the Spanner RPCs) which is provided by the Spanner client and a user provides `GcpManagedChannelOptions` which configures other aspects like metrics and resiliency options.

Reply:
I think changing `maxConcurrentStreamsLowWatermark` to 0 will answer my last questions / worries. Other than that, this seems good to me.

Reply:
Changed low watermark to zero.